### PR TITLE
[SPARK-49695][SQL] Postgres fix xor push-down

### PR DESCRIPTION
**Pull request by andrej-db**
### What changes were proposed in this pull request?
This PR fixes the pushdown of ^ operator (XOR operator) for Postgres. Those two databases use this as exponent, rather then bitwise xor.

Fix is consisted of overriding the SQLExpressionBuilder to replace the '^' character with '#'.
### Why are the changes needed?
Result is incorrect.

### Does this PR introduce _any_ user-facing change?
Yes. The user will now have a proper translation of the ^ operator.

### Was this patch authored or co-authored using generative AI tooling?
No.

### CODE REVIEW

**Review by PetarVasiljevic-DB — APPROVED** [review]

**Review by urosstan-db — APPROVED** [review]

**Comment by urosstan-db** [comment]
Can we add some tests?

**Review by MaxGekk — CHANGES_REQUESTED** [review]
It would be nice if you add an integration test, for instance here `*PostgresExpressionPushdownSuite*`

**Comment by andrej-db** [comment]
I wanted to add tests, but didn't know where to put them... Don't know how oss tests postgre.

**Comment by urosstan-db** [comment]
> I wanted to add tests, but didn't know where to put them... Don't know how oss tests postgre.

`V2JDBCTest` is base class, and there is derived class for Postgres `PostgresIntegrationSuite`

**Review by dongjoon-hyun — COMMENTED** [review]
Thank you, @andrej-db . +1 for adding a test case definitely.

cc @huaxingao , too.

**Review by RaleSapic — APPROVED** [review]

**Comment by andrej-db** [comment]
Added the test, let me know if this is in order.

**Review by urosstan-db — APPROVED** [review]

File: connector/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/v2/V2JDBCTest.scala [hunk 1]
```diff
@@ -986,4 +986,13 @@ private[v2] trait V2JDBCTest extends SharedSparkSession with DockerIntegrationFu
   test("scan with filter push-down with date time functions") {
     testDatetime(s"$catalogAndNamespace.${caseConvert("datetime")}")
   }
+
+  test("xor operator push-down") {
```
Comment (urosstan-db) [inline connector/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/v2/V2JDBCTest.scala:990]: Maybe you can do explain formatted and check whether string contains `"id" # 3`, that will add a little bit of robustness.
Another thing we can do is to make unit test, and just invoke compilation of XOR expression and check whether `col # constant` is result of compilation.

**Review by urosstan-db — APPROVED** [review]
Approving latest iteration (with tests)

File: connector/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/v2/V2JDBCTest.scala (hunk 1 above)
Comment (andrej-db) [reply connector/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/v2/V2JDBCTest.scala:990]: neat, I like it

**Comment by andrej-db** [comment]
jenkins trigger: Spark-Integration-Snowflake-Pr

**Review by MaxGekk — CHANGES_REQUESTED** [review]
Please, fix your test:
```
[info] - SPARK-49695: Postgres fix xor push-down *** FAILED *** (10 milliseconds)
[info]   org.apache.spark.sql.catalyst.ExtendedAnalysisException: [TABLE_OR_VIEW_NOT_FOUND] The table or view `bar` cannot be found. Verify the spelling and correctness of the schema and catalog.
```

File: sql/core/src/main/scala/org/apache/spark/sql/jdbc/PostgresDialect.scala [hunk 2]
```diff
@@ -404,4 +404,26 @@ private case class PostgresDialect()
       case _ =>
     }
   }
+
+  override def compileExpression(expr: Expression): Option[String] = {
+    val builder = new PostgresSQLBuilder()
+    try {
```
Comment (milastdbx) [inline sql/core/src/main/scala/org/apache/spark/sql/jdbc/PostgresDialect.scala:410]: perhaps its better to only override visitBinaryArithmetics?

We have similar problem with some of the functions and we use `dialectFunctionName` to translate from Spark to local dialect

**Review by milastdbx — COMMENTED** [review]
Consider refactoring this

**Review by milastdbx — APPROVED** [review]

File: connector/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/v2/PostgresIntegrationSuite.scala [hunk 3]
```diff
@@ -230,6 +230,14 @@ class PostgresIntegrationSuite extends DockerJDBCIntegrationV2Suite with V2JDBCT
     }
   }
 
+  test("SPARK-49695: Postgres fix xor push-down") {
+    val df = spark.sql("select id, name from employee where id ^ 6 = 0")
```
Comment (urosstan-db) [inline connector/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/v2/PostgresIntegrationSuite.scala:234]:
```suggestion
    val df = spark.sql("select id, name from employee where id ^ 6 = 0")
    assert(!df.queryExecution.optimizedPlan.exists { plan =>
      plan.isInstanceOf[LogicalPlan]
    })
```

We should assert xor is actually pushed down here

File: sql/core/src/main/scala/org/apache/spark/sql/jdbc/PostgresDialect.scala (hunk 2 above)
Comment (urosstan-db) [reply sql/core/src/main/scala/org/apache/spark/sql/jdbc/PostgresDialect.scala:410]: +1 here, let's override last possible method in chain of execution

**Review by urosstan-db — APPROVED** [review]
LGTM, but please make test more assertive.

File: sql/core/src/main/scala/org/apache/spark/sql/jdbc/PostgresDialect.scala (hunk 2 above)
Comment (andrej-db) [reply sql/core/src/main/scala/org/apache/spark/sql/jdbc/PostgresDialect.scala]: done

**Review by MaxGekk — CHANGES_REQUESTED** [review]
@andrej-db Could you fix builds and retrigger intergration tests:
```
[error] /home/runner/work/spark/spark/connector/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/v2/PostgresIntegrationSuite.scala:237:25: not found: type Filter
[error]       plan.isInstanceOf[Filter]
[error]                         ^
[error] one error found
```

File: connector/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/v2/PostgresIntegrationSuite.scala [hunk 4]
```diff
@@ -252,6 +253,17 @@ class PostgresIntegrationSuite extends DockerJDBCIntegrationV2Suite with V2JDBCT
...
 
+  test("SPARK-49695: Postgres fix xor push-down") {
+    val df = spark.sql(s"select dept, name from $catalogName.employee where dept ^ 6 = 0")
+    val rows = df.collect()
+    assert(df.queryExecution.sparkPlan.collectFirst {
+      case f: FilterExec => f
+    }.isEmpty)
+    assert(rows.length == 1)
+    assert(rows(0).getInt(0) === 6)
+    assert(rows(0).getString(1) === "jen")
```
Comment (MaxGekk) [inline connector/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/v2/PostgresIntegrationSuite.scala:264]: Could you use `checkAnswer`, please.

File: connector/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/v2/PostgresIntegrationSuite.scala [hunk 5]
```diff
@@ -252,6 +253,17 @@ class PostgresIntegrationSuite extends DockerJDBCIntegrationV2Suite with V2JDBCT
     }
   }
 
+  test("SPARK-49695: Postgres fix xor push-down") {
+    val df = spark.sql(s"select dept, name from $catalogName.employee where dept ^ 6 = 0")
+    val rows = df.collect()
+    assert(df.queryExecution.sparkPlan.collectFirst {
+      case f: FilterExec => f
+    }.isEmpty)
```
Comment (MaxGekk) [inline connector/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/v2/PostgresIntegrationSuite.scala:261]:
```suggestion
    assert(!df.queryExecution.sparkPlan.exists(_.isInstanceOf[FilterExec]))
```

File: connector/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/v2/PostgresIntegrationSuite.scala (hunk 4 above)
Comment (andrej-db) [reply connector/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/v2/PostgresIntegrationSuite.scala:262]: This file has mainly these asserts, wanted to make it in line with other tests

**Review by MaxGekk — APPROVED** [review]

**Comment by MaxGekk** [comment]
+1, LGTM. Merging to master/3.5/3.4.
Thank you, @andrej-db and @urosstan-db @RaleSapic @PetarVasiljevic-DB @milastdbx @dongjoon-hyun for review.

**Comment by MaxGekk** [comment]
@andrej-db Could you open separate PRs for 3.4 and 3.5 because your changes cause conflicts in the branches.

**Comment by dongjoon-hyun** [comment]
To @MaxGekk and @andrej-db , Apache Spark 3.4 reached the End-Of-Support. 

Only `branch-3.5` is open for backporting. 

Also, cc @LuciferYang since he is the release manager for Apache Spark 3.5.4.

**Comment by LuciferYang** [comment]
Thanks @dongjoon-hyun

**Pull request closed without merging**